### ISSUE TITLE
[Bug] SamlSecurityTokenHandler and Saml2SecurityTokenHandler does not fetch configuration from ConfigurationManager on ValidateTokenAsync

### ISSUE THREAD

**Issue by Hakon** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#2406]
**Which version of Microsoft.IdentityModel are you using?**

Microsoft.AspNetCore.Authentication.WsFederation 8.0.0
Microsoft.IdentityModel.Protocols.WsFederation 7.0.3
Microsoft.IdentityModel.Tokens.Saml 7.0.3

**Where is the issue?**
 * [ ] M.IM.JsonWebTokens
 * [ ] M.IM.KeyVaultExtensions
 * [ ] M.IM.Logging
 * [ ] M.IM.ManagedKeyVaultSecurityKey
 * [ ] M.IM.Protocols
 * [ ] M.IM.Protocols.OpenIdConnect
 * [ ] M.IM.Protocols.SignedHttpRequest
 * [ ] M.IM.Protocols.WsFederation
 * [ ] M.IM.TestExtensions
 * [ ] M.IM.Tokens
 * [X] M.IM.Tokens.Saml
 * [ ] M.IM.Validators
 * [ ] M.IM.Xml
 * [ ] S.IM.Tokens.Jwt
 * Other (please describe)
 * [x] M.AspNetCore.Authentication.WsFederation

**Is this a new or an existing app?**
The app is in production and I have upgraded to a new version of Microsoft.AspNetCore.Authentication.WsFederation and Microsoft.IdentityModel.Tokens.Saml in development.

**Repro**

```csharp
var wsfedmessage = new WsFederationMessage();
wsfedmessage.Wa = "wsignin1.0";
wsfedmessage.Wresult = @"<your wresult here>";
wsfedmessage.Wctx =
    "your wctx here";

var token = wsfedmessage.GetToken();
var handler = new SamlSecurityTokenHandler();

var validationParameters = new TokenValidationParameters();
validationParameters.ConfigurationManager = new ConfigurationManager<WsFederationConfiguration>(
    "https://<your wsfederation host here>/FederationMetadata/2007-06/FederationMetadata.xml",
    new WsFederationConfigurationRetriever());

var result = await handler.ValidateTokenAsync(token, validationParameters);
if (!result.IsValid)
{
    throw result.Exception;
}
```

**Expected behavior**
I expect SamlSecurityTokenHandler to fetch configuration using the ConfigurationManager and result to be valid.  Microsoft.AspNetCore.Authentication.WsFederation expects the handler to fetch Configuration using the ConfigurationManager as it is done in the JsonWebTokenHandler.

**Actual behavior**
result.IsValid == false with the exception "IDX10204: Unable to validate issuer. validationParameters.ValidIssuer is null or whitespace AND validationParameters.ValidIssuers is null or empty.".
This is because the configuration is not fetched.

**Possible solution**
Fetch the current configuration at the start of ValidateTokenAsync and pass this with a new TokenValidationParameters to ValidateToken. I don't know the solution for the non-async ValidateToken as GetBaseConfigurationAsync is async.

**Comment by Hakon** [comment]
I have worked around this issue with the following handler:
```
public class CustomSamlSecurityTokenHandler : SamlSecurityTokenHandler
{
    public override async Task<TokenValidationResult> ValidateTokenAsync(string token, TokenValidationParameters validationParameters)
    {

        var configuration = await validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false);
        var issuers = new[] { configuration.Issuer };
        validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? issuers : validationParameters.ValidIssuers.Concat(issuers));
        validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(configuration.SigningKeys));

        return await base.ValidateTokenAsync(token, validationParameters);
    }
}
```

**Comment by MichaelBuehler** [comment]
We get this issue as well - I have a question to the Workaround by Hakon:  Can you share the code how add CustomSamlSecurityTokenHandler?

**Comment by fabiano-ferreira** [comment]
Yes @MichaelBuehler!

I also would like to know how to use the method ValidateTokenAsync.

**Comment by Hakon** [comment]
In your AddWsFederation options add the following: 
```C#
                options.TokenHandlers.Clear();
                options.TokenHandlers.Add(new CustomSamlSecurityTokenHandler());
```
Only do the Clear(); if you know you are not going to encounter Saml2 tokens or jwt.

**Comment by fabiano-ferreira** [comment]
> ```cs
>  options.TokenHandlers.Clear();
>                 options.TokenHandlers.Add(new CustomSamlSecurityTokenHandler());
> ```

Hi @Hakon!

I am trying to use your workaround and now I am getting a different error:

`XmlReadException: IDX30011: Unable to read XML. Expecting XmlReader to be at ns.element: '[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.].[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]', found: '[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.].[PII of type 'System.String' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'.
Microsoft.IdentityModel.Xml.XmlUtil.CheckReaderOnEntry(XmlReader reader, string element, string namespace)
Microsoft.IdentityModel.Tokens.Saml.SamlSerializer.ReadAssertion(XmlReader reader)
Microsoft.IdentityModel.Tokens.Saml.SamlSecurityTokenHandler.ReadSamlToken(XmlReader reader)
Microsoft.IdentityModel.Tokens.Saml.SamlSecurityTokenHandler.ReadSamlToken(string token)
Microsoft.IdentityModel.Tokens.Saml.SamlSecurityTokenHandler.ValidateSignature(string token, TokenValidationParameters validationParameters)
Microsoft.IdentityModel.Tokens.Saml.SamlSecurityTokenHandler.ValidateToken(string token, TokenValidationParameters validationParameters, out SecurityToken validatedToken)
Microsoft.IdentityModel.Tokens.Saml.SamlSecurityTokenHandler.ValidateTokenAsync(string token, TokenValidationParameters validationParameters)
`

Do you have any idea about what can be happening?

**Comment by Hakon** [comment]
Hard to say without the PII-strings. Do you know if that's a saml1 or saml2 token?

**Comment by MichaelBuehler** [comment]
Hi @Hakon
Your fix worked at least for me, thank you!

For the record, I am using the dll versions mentioned above (currently the latest), issue appeared after upgrade to .NET 8:
Microsoft.AspNetCore.Authentication.WsFederation 8.0.0
Microsoft.IdentityModel.Tokens.Saml 7.0.3  (part of nuget Microsoft.IdentityModel.Tokens 7.0.3 (not sure why there is no 8.0.0)

**Comment by fabiano-ferreira** [comment]
> Hard to say without the PII-strings. Do you know if that's a saml1 or saml2 token?

Hi @Hakon!

This is the answer that I had:

`An unhandled exception occurred while processing the request.
XmlReadException: IDX30011: Unable to read XML. Expecting XmlReader to be at ns.element: 'urn:oasis:names:tc:SAML:1.0:assertion.Assertion', found: 'urn:oasis:names:tc:SAML:2.0:assertion.Assertion'.
Microsoft.IdentityModel.Xml.XmlUtil.CheckReaderOnEntry(XmlReader reader, string element, string namespace)

SecurityTokenException: No token validator or token handler was found for the given token.
Microsoft.AspNetCore.Authentication.WsFederation.WsFederationHandler.HandleRemoteAuthenticateAsync()

AuthenticationFailureException: An error was encountered while handling the remote login.
Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler<TOptions>.HandleRequestAsync()`

**Comment by OttoG** [comment]
> I have worked around this issue with the following handler:
> 
> ```c#
> public class CustomSamlSecurityTokenHandler : SamlSecurityTokenHandler
> {
>     public override async Task<TokenValidationResult> ValidateTokenAsync(string token, TokenValidationParameters validationParameters)
>     {
> 
>         var configuration = await validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false);
>         var issuers = new[] { configuration.Issuer };
>         validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? issuers : validationParameters.ValidIssuers.Concat(issuers));
>         validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(configuration.SigningKeys));
> 
>         return await base.ValidateTokenAsync(token, validationParameters);
>     }
> }
> ```

Nice workaround! I did not realize that the configuration read from the `MetadataAddress` set in the options was readily available to the token validation method, so that it could be picked up by a subclass that sets the missing values and then calls the base method.

I came up with a workaround that will be less ideal, as it does not benefit from the built-in logic for automatically retrying configuration fetching. But it is simpler in a way and may therefore be useful to anyone who wants to verify that they suffer from the same issue, before they spend time on possibly having to troubleshoot an implementation of the above workaround.

Simply set the following two parameters in the same place where `options.MetadataAddress`, `options.WtRealm`, etc, are set:

```c#
options.TokenValidationParameters.ValidIssuer = "[URL-style string found in the entityID attribute of the root element of the metadata XML]";
options.TokenValidationParameters.IssuerSigningKey = new X509SecurityKey(new X509Certificate2(Convert.FromBase64String("[Base64 string found in the KeyInfo element inside a ds:Signature element of the metadata XML]");
```

If there are multiple valid issuers or keys, use the `ValidIssuers` and `IssuerSigningKeys` parameters instead.

I then went on to reading the settings directly from the metadata XML instead (using `new WsFederationMetadataSerializer().ReadMetadata(System.Xml.XmlReader.Create(stream))`), but I would recommend trying to get @Hakon’s workaround to work, if possible. Though of course, if it is hard to find the values manually in the XML, this retrieves them in a structured format (the `Issuer` and `SigningKeys` properties) to use when experimenting.

**Comment by OttoG** [comment]
> This is the answer that I had:
> 
> `An unhandled exception occurred while processing the request. XmlReadException: IDX30011: Unable to read XML. Expecting XmlReader to be at ns.element: 'urn:oasis:names:tc:SAML:1.0:assertion.Assertion', found: 'urn:oasis:names:tc:SAML:2.0:assertion.Assertion'. Microsoft.IdentityModel.Xml.XmlUtil.CheckReaderOnEntry(XmlReader reader, string element, string namespace)
> 
> SecurityTokenException: No token validator or token handler was found for the given token. Microsoft.AspNetCore.Authentication.WsFederation.WsFederationHandler.HandleRemoteAuthenticateAsync()
> 
> AuthenticationFailureException: An error was encountered while handling the remote login. Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler.HandleRequestAsync()`

I had similar error messages until I came up with a workaround. (Or sometimes quite different error messages, depending on experimentation with various parameters and removal of token handlers.) The messages make it look like if it all has something to do with SAML 1.0 or SAML 2.0 versions of the token, but in fact, it was missing configuration data that was the problem.

But the versions were mentioned the other way around for me – “expecting …2.0…, found …1.0…” – which makes me think that perhaps your authentication token @fabiano-ferreira is in SAML **2.0** format. In that case, you should subclass the `Saml2SecurityTokenHandler` instead of the `SamlSecurityTokenHandler` used in @Hakon’s example. That would explain why his workaround does not (yet) work for you.

**Comment by fabiano-ferreira** [comment]
> > This is the answer that I had:
> > `An unhandled exception occurred while processing the request. XmlReadException: IDX30011: Unable to read XML. Expecting XmlReader to be at ns.element: 'urn:oasis:names:tc:SAML:1.0:assertion.Assertion', found: 'urn:oasis:names:tc:SAML:2.0:assertion.Assertion'. Microsoft.IdentityModel.Xml.XmlUtil.CheckReaderOnEntry(XmlReader reader, string element, string namespace)
> > SecurityTokenException: No token validator or token handler was found for the given token. Microsoft.AspNetCore.Authentication.WsFederation.WsFederationHandler.HandleRemoteAuthenticateAsync()
> > AuthenticationFailureException: An error was encountered while handling the remote login. Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler.HandleRequestAsync()`
> 
> I had similar error messages until I came up with a workaround. (Or sometimes quite different error messages, depending on experimentation with various parameters and removal of token handlers.) The messages make it look like if it all has something to do with SAML 1.0 or SAML 2.0 versions of the token, but in fact, it was missing configuration data that was the problem.
> 
> But the versions were mentioned the other way around for me – “expecting …2.0…, found …1.0…” – which makes me think that perhaps your authentication token @fabiano-ferreira is in SAML **2.0** format. In that case, you should subclass the `Saml2SecurityTokenHandler` instead of the `SamlSecurityTokenHandler` used in @Hakon’s example. That would explain why his workaround does not (yet) work for you.

Thank you @Hakon and @OttoG !!

After the problem was really the SAML version. Now after changing to this code, it works for me:

`public class CustomSaml2SecurityTokenHandler : Saml2SecurityTokenHandler
{
    public override async Task<TokenValidationResult> ValidateTokenAsync(string token, TokenValidationParameters validationParameters)
    {

        var configuration = await validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false);
        var issuers = new[] { configuration.Issuer };
        validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? issuers : validationParameters.ValidIssuers.Concat(issuers));
        validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(configuration.SigningKeys));

        return await base.ValidateTokenAsync(token, validationParameters);
    }
}`

**Comment by brentschmaltz** [comment]
@Hakon @OttoG @MichaelBuehler @fabiano-ferreira @RamonPage This is clearly a test hole that needs to be plugged.
The workaround suggested by @OttoG is similar to what asp.net did internally prior to net8.

It is **important** to clone TokenValidationParameters before calling base.ValidateTokenAsync since TokenValidationParameters can persist between http requests.

**Comment by OttoG** [comment]
Thank you @brentschmaltz for the warning. This could easily be very important in some scenarios.

However, we have a setup where we only ever expect to get one single configuration in a given environment, there is already a condition in place in our code to avoid changing the parameters if the valid issuer and issuer signing keys are already set (see below, to avoid calling ValidateTokenAsync after the problem has been fixed in .NET), and the assignments are of reference types and are thus atomic. The fact that the changed parameters can persist then effectively becomes a small optimization.

Of course, if the TokenValidationParameters objects persist for very long times, it could become a problem if the metadata sent by the authentication server eventually changes. But that, too, is unlikely to happen in our case.

I mention this in case anyone fears that they have deployed workarounds that may cause security bugs but actually has environments similar to ours.

```c#
public class FixedSamlSecurityTokenHandler : SamlSecurityTokenHandler
{
    public override async Task<TokenValidationResult> ValidateTokenAsync(string token, TokenValidationParameters parameters)
    {
        if ((string.IsNullOrEmpty(parameters.ValidIssuer) && parameters.ValidIssuers?.Any() != true) || parameters.IssuerSigningKeys?.Any() != true)
        {
            var configuration = await parameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None);
            parameters.ValidIssuer = configuration.Issuer;
            parameters.IssuerSigningKeys = configuration.SigningKeys;
        }
        return await base.ValidateTokenAsync(token, parameters);
    }
}
```

**Comment by Magnus-K-work** [comment]
I was upgrading an old project to .Net 8 and a similar error message was returned during a login sequence.
After comparing nuget references with a similar project that used .Net7, I downgraded Microsoft.AspNetCore.Authentication.WsFederation to latest 7.0.x version in my .Net 8 project and got the login to work again.
How can I troubleshoot what is going on in the 8.0.x version of Microsoft.AspNetCore.Authentication.WsFederation when the error occurs?

**Comment by brentschmaltz** [comment]
@OttoG We coordinated with to help with stale credentials, where we throw the exception SecurityTokenSignatureKeyNotFoundException, which occurs when the token being validated has a keyid, but does not match any of the keys available.

**Comment by jesslilly** [comment]
Confirming that `CustomSamlSecurityTokenHandler ` is working wonderfully.  Thanks a TON!

**Comment by brentschmaltz** [comment]
@Hakon this is a new feature.
I marked the issue as such.

**Comment by Hakon** [comment]
What is the new feature here? This is clearly a regression as code that previously worked before does not work now.

**Comment by westin-m** [comment]
@Hakon The fix for this has been included in the 7.4.1 release

**Comment by pcibraro** [comment]
@westin-m Microsoft.AspNetCore.Authentication.WsFederation has not been updated to reference 7.4.1. The latest version 8.0.4 still references to 7.1.2. Do you know if there is any plan to get that updated ? Thanks

**Comment by westin-m** [comment]
@pcibraro I do not know if such plans exist. Please bring your question to [the aspnetcore repo](https://github.com/dotnet/aspnetcore) where I think they will be able to answer your question.

**Comment by philipstratford** [comment]
> I was upgrading an old project to .Net 8 and a similar error message was returned during a login sequence. After comparing nuget references with a similar project that used .Net7, I downgraded Microsoft.AspNetCore.Authentication.WsFederation to latest 7.0.x version in my .Net 8 project and got the login to work again. 

I've implemented this workaround, such as it is, from @Magnus-K-work, and it's worked for me. I can't say I've completely understood why this issue has cropped up as a result of me migrating a web app project from .NET 7 to .NET 8 (hosted on our servers, nothing to do with Azure, though we do use ADFS for authentication), but I'm hoping that at some point in the future the WSFederation pacakge might be updated to allow us to use the 8.x.x version without the need for a custom token handler.

> this is a new feature

Sorry, what's the feature..? I can't decide if this comment is a joke or not (I think not), but I also can't see how adding a requirement for devs to write their own custom token handler is a feature?

**Comment by kobudzik** [comment]
It still doesn't work without the CustomSamlSecurityTokenHandler workaround, at AspNetCore.Authentication.WsFederation v8.0.8.

 I was updating Identity server and .Net versions, it was tricky to find what is happening.

**Comment by marco-innovcast** [comment]
Bumping Microsoft.IdentityModel.Tokens.Saml to the latest (8.3.0) after bumping to .NET 8 solves that issue for me.
I do get into a redirect loop afterwards but that is a new kind of problem not related to this 😅

**Comment by SirCypkowskyy** [comment]
@marco-innovcast I am also struggling with the redirect loop error :( If you find a solution to this problem, I would be more than thankful for getting it.

From what I understand, it may have something to do with [https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/troubleshooting/ad-fs-tshoot-loop](https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/troubleshooting/ad-fs-tshoot-loop), but I have no idea how it could be fixed.

**Comment by marco-innovcast** [comment]
@SirCypkowskyy For ADFS I've lowered the pkg Microsoft.IdentityModel.Tokens.Saml to 7.7.0 and it worked fine.
Now I'm having the redirect loop on AAD.

**Comment by SirCypkowskyy** [comment]
> @SirCypkowskyy For ADFS I've lowered the pkg Microsoft.IdentityModel.Tokens.Saml to 7.7.0 and it worked fine. Now I'm having the redirect loop on AAD.

I have a similar problem. We use Traefik for proxying, and due to it, I cannot run my app in HTTPS. I made it (ws-federation) work under HTTPS, but it does not work if I try to run it under proxy (I even tried forcefully transform the http addresses to https, but it still failed or later on it fell under redirect loop).

I also tried manually configuring it with OIDC, and it even partially worked (tho I had problems with validating the token in other microservices, under YARP Gateway), but then I broke it again :(

I would really love to find a solution to this, and I'll keep you informed if you want about my progress.
